### PR TITLE
RAC-5948   Add Manifest functionality as a common library

### PR DESCRIPTION
**Pull request by panpan0000**
Usage/Unit-Test attached at the end of this Groovy file.

Test done on Sandbox:  http://10.62.**.**5:8080/job/test_manifest/70

Functions:
- generateManifest
-publishManifest
-downloadManifest
-stashManifest
-unstashManifest
-checkoutAccordingToManifest
-checkoutTargetRepo

Reviewer:
@PengTian0  @changev

### CODE REVIEW

File: src/pipeline/common/Manifest.groovy [hunk 1]
```diff
@@ -1,38 +1,248 @@
...
+
+         sh """#!/bin/bash -ex
+            ${on_build_config_dir}/build-release-tools/HWIMO-BUILD ${on_build_config_dir}/build-release-tools/application/generate_manifest.py \
+            --branch $branch \
+            --date $date \
+            --timezone $timezone \
+            --builddir $target_dir \
+            --force \
+            --jobs 8
+         """
```
Comment (PengTian0) [inline src/pipeline/common/Manifest.groovy]: indent

File: src/pipeline/common/Manifest.groovy [hunk 2]
```diff
@@ -1,38 +1,248 @@
...
+        // note: in sh """  """, the $ is a groovy resvered key. so you need to use \$ if you want shell logic like "echo ${shell_var}"
+        fname = sh (
```
Comment (PengTian0) [inline src/pipeline/common/Manifest.groovy]: How about fpath?

Comment (panpan0000) [reply src/pipeline/common/Manifest.groovy]: good idea, fixed

File: src/pipeline/common/Manifest.groovy (hunk 1 above)
Comment (panpan0000) [reply src/pipeline/common/Manifest.groovy]: fixed

**Comment by PengTian0** [comment]
Could you please add function:
```
generateManifestFromPR
generateManifestFromLocalRepos (generate_manifest.py can generate a manifest file according to the latest commit of repositories under local directory)（https://github.com/RackHD/on-build-config/blob/72722efd0a0ffbe85e0af7199f56c5c68ab53129/jobs/SprintRelease/update_manifest.sh#L3）
getRepoUnderTest(manifest_path) (https://github.com/RackHD/on-build-config/blob/b57b51d30208390e9822ef2de6a491a486569198/jobs/UnitTest/PRGateUnitTest.groovy#L35)
```

**Comment by panpan0000** [comment]
@PengTian0 
3 new functions just added, please review again, all tested.
````
generateManifestFromPR
generateManifestFromLocalRepos
getReposNeedTest
```

**Comment by PengTian0** [comment]
👍

**Review by iceiilin — APPROVED** [review]

**Pull request merged**